src/include/rbtree.h

```diff
@@ -1,17 +1,173 @@
+ * @tree: Pointer to the red-black tree from which data is to be retrieved.
+ * @addr: The memory address used as a key to find the corresponding node in the tree.
+ *
+ * This function searches the red-black tree for a node that corresponds to the specified address.
+ * Return: The data associated with the address if found; otherwise, 0.
+ */
 uint32_t rb_get(rbtree_t* tree, uintptr_t addr);
+
+/** 
+ * rb_get_end() - Searches for a node within a specified address range in a red-black tree and retrieves its data and end address.
+ * @tree: Pointer to the red-black tree to be searched.
+ * @addr: The address to search for within the nodes of the red-black tree.
+ * @val: Pointer to store the data of the node that contains the address if found.
+ * @end: Pointer to store the end address of the node that contains the address, or the start of the next node if not found, or UINTPTR_MAX if no next node exists.
+ *
+ * This function traverses the red-black tree starting from the root, searching for a node where the 'addr' falls between the node's start and end addresses (exclusive of end).
+ * If such a node is found, the function stores the node's data in '*val' and the node's end address in '*end', then returns 1 to indicate success.
+ * If no such node is found, the function sets '*val' to 0 and '*end' to the start address of the next node in the tree structure or to UINTPTR_MAX if there is no subsequent node.
+ * Return: 1 if a node containing the address is found, otherwise 0.
+ */
 int rb_get_end(rbtree_t* tree, uintptr_t addr, uint32_t* val, uintptr_t* end);
+
+/**
+ * rb_set() - Set an address range in a red-black tree.
+ * @tree: Pointer to the red-black tree where the address range will be set.
+ * @start: The starting address of the range to be set.
+ * @end: The ending address of the range to be set.
+ * @data: The data value to associate with the address range.
+ *
+ * This function adds a new address range with associated data to the red-black tree. 
+ * If the range overlaps with existing nodes, it intelligently merges or modifies nodes to maintain non-overlapping, contiguous ranges.
+ * This includes extending, splitting, or merging nodes based on the overlap conditions and data consistency. 
+ * It handles multiple edge cases:
+ * 1. Overlap with same data: The existing node will be extended.
+ * 2. Overlap with different data: The existing node will split into two or three nodes depending on the position of the new address range, 
+ *    and the overlapped part will be overwritten with the data of the new address range.
+ * 
+ * (case 1)
+ *       +---------+---------+---------+---------+---------+
+ *       |                   data A                        |
+ *       +---------+---------+---------+---------+---------+
+ *                  ^.....       overlap      ....^
+ *                  +---------+---------+---------+
+ *                  |            data B           |
+ *                  +---------+---------+---------+
```

Comment:
This is nitpicking, but this is misaligned. Also, the `overlap` is not centered, even though it could be (there are five dits and seven spaces on the left, six spaces and four dots on the right).
```suggestion
 *                 ^.....      overlap      .....^
 *                 +---------+---------+---------+
 *                 |            data B           |
 *                 +---------+---------+---------+
```